### PR TITLE
[pallas] Add more documentation and tests for BlockSpec.

### PR DESCRIPTION
**Pull request by gnecula**
This PR deals with the default values for the parameters of the `BlockSpec` constructor, and the mapped block dimensions.

Fix a bug where previously a missing block_shape while the index_map was present was resulting in a crash.

### CODE REVIEW

File: docs/pallas/grid_blockspec.md [hunk 1]
```diff
@@ -33,7 +33,10 @@ for i in range(n):
 This generalizes to any tuple of integers (a length `d` grid will correspond
 to `d` nested loops).
 The kernel is executed as many times
-as `prod(grid)`. Each of these invocations is referred to as a "program".
+as `prod(grid)`.
+The default grid value `None` stands for `()`, and results in one
```
Comment (apaszke) [inline docs/pallas/grid_blockspec.md:37]: Hmm can't we change the default to `()`? There's no reason for a user to ever pass in `None` and I would not mention that in the docs

File: docs/pallas/grid_blockspec.md [hunk 2]
```diff
@@ -33,7 +33,10 @@ for i in range(n):
...
+kernel invocation.
+Each of these invocations is referred to as a "program".
```
Comment (apaszke) [inline docs/pallas/grid_blockspec.md:39]: invocation -> instance?

File: docs/pallas/grid_blockspec.md [hunk 3]
```diff
@@ -233,3 +236,46 @@ See [the Pallas TPU documentation](https://jax.readthedocs.io/en/latest/pallas/t
...
+...                grid=(3, 2), interpret=True)()
+Array([[ 0,  0, 10, 10],
+       [ 1,  1, 11, 11],
+       [ 2,  2, 12, 12]], dtype=int32)
+
+```
+
+When we construct a `BlockSpec` we can use the value `None` for the
+`block_shape` parameter, in which case the shape of the overall array
+is used. And if we use the value `None` for the `index_map` parameter
```
Comment (apaszke) [inline docs/pallas/grid_blockspec.md:262]: the shape of the entire array is used -> the entire array is used? It's not just the shape, but really the slice is the whole array

File: docs/pallas/grid_blockspec.md [hunk 4]
```diff
@@ -233,3 +236,46 @@ See [the Pallas TPU documentation](https://jax.readthedocs.io/en/latest/pallas/t
...
+then a default index map function that returns a tuple of zeros is
+used: `index_map=lambda *invocation_indices: (0,) * len(block_shape)`.
```
Comment (apaszke) [inline docs/pallas/grid_blockspec.md:264]: Oh wow I had no idea about this. I don't think this is even desirable? It makes no sense to specify a BlockSpec without an index function. Can we not document this behavior and perhaps restrict it?

File: jax/_src/pallas/core.py [hunk 5]
```diff
@@ -310,9 +313,11 @@ def _convert_block_spec_to_block_mapping(
     return None
   if block_spec.index_map is None:
     compute_index = lambda *args, **kwargs: (0,) * len(aval.shape)
-    block_shape = aval.shape
   else:
     compute_index = block_spec.compute_index
+  if block_spec.block_shape is None:
+    block_shape = aval.shape
```
Comment (apaszke) [inline jax/_src/pallas/core.py:319]: That doesn't make a lot of sense either. A block spec should always contain both fields

File: jax/_src/pallas/pallas_call.py [hunk 6]
```diff
@@ -1034,12 +1034,14 @@ def pallas_call(
       See details at :ref:`pallas_grid`.
     in_specs: a PyTree of :class:`jax.experimental.pallas.BlockSpec` with
       a structure matching that of the positional arguments.
+      The default value for ``in_specs`` specifies the whole array for all
+      inputs, e.g., as ``pl.BlockSpec(x.shape, lambda *indices: zeros)``.
```
Comment (apaszke) [inline jax/_src/pallas/pallas_call.py:1038]: Perhaps say `lambda *_: (0,) * x.ndim`? `zeros` is undefined

File: jax/_src/pallas/pallas_call.py [hunk 7]
```diff
@@ -1063,6 +1065,9 @@ def pallas_call(
   if grid_spec is None:
     grid_spec = GridSpec(grid, in_specs, out_specs)
   grid_spec, dynamic_grid_bounds = grid_spec.unzip_dynamic_grid_bounds()
+  # TODO(necula): this canonicalization may be convenient for some usage
+  # but it is lossy, because it prevents expressing functions that return
+  # lists.
```
Comment (apaszke) [inline jax/_src/pallas/pallas_call.py:1070]: I think it's common for many JAX functions to do it?

File: tests/pallas/pallas_test.py [hunk 8]
```diff
@@ -253,6 +253,114 @@ def test_pallas_call_out_shape_is_list(self):
     # TODO(necula): we normalize out_shape to a tuple, we shouldn't.
     self.assertIsInstance(res, tuple)
 
+  @jtu.skip_on_devices("gpu")  # TODO: RET_CHECK failure
```
Comment (apaszke) [inline tests/pallas/pallas_test.py:256]: Is this TODO resolved?

**Review by apaszke — CHANGES_REQUESTED** [review]
I think BlockSpecs should always have both fields.

File: tests/pallas/pallas_test.py (hunk 8 above)
Comment (gnecula) [reply tests/pallas/pallas_test.py:256]: No, it is not resolved, but I see that as separate work.

File: docs/pallas/grid_blockspec.md (hunk 1 above)
Comment (gnecula) [reply docs/pallas/grid_blockspec.md:37]: Indeed, I wanted to ask about that. I'd like to do that separately, because it changes the API.

File: docs/pallas/grid_blockspec.md (hunk 4 above)
Comment (gnecula) [reply docs/pallas/grid_blockspec.md:264]: I agree that there is too much optionality in the API. We should be allowed to skip BlockSpec, but if we have them they should include the block_shape and index_map. If there is agreement about that, I would do it in a different PR, because it changes the API.

Comment (gnecula) [reply docs/pallas/grid_blockspec.md:264]: @sharadmv Agree with less optionality?

File: jax/_src/pallas/pallas_call.py (hunk 7 above)
Comment (gnecula) [reply jax/_src/pallas/pallas_call.py:1070]: Is it? I think only for the inputs, because the inputs are always a tuple, but I don't think we do this for outputs, or at least not for the cases where we determine whether the output is a tuple or a list. (E.g., in `pjit` the output container is determined by the user when they write the function, and the `out_shardings` simply annotation the outputs; but `out_shape` actually determines the container.)

See https://github.com/search?q=repo%3Agoogle%2Fjax%20https%3A%2F%2Fgithub.com%2Fgoogle%2Fjax%2Fissues%2F2367&type=code

File: docs/pallas/grid_blockspec.md (hunk 2 above)
Comment (gnecula) [reply docs/pallas/grid_blockspec.md:39]: I tried to be consistent and the whole file uses "kernel invocation". Initially there was a mix of kernel invocation and iteration. Do you think that "instance" is better than "invocation" (to use throughout)?

**Comment by gnecula** [comment]
> I think BlockSpecs should always have both fields.

I agree that there is too much optionality in the API. We should be allowed to skip BlockSpec, but if we have them they should include the block_shape and index_map.

@sharadmv @superbobry  do you agree with less optionality?

 If there is agreement about that, I would do it in a different PR, because it changes the API.
